File: 1525 Number of Good Ways to Split a String/1525 Number of Good Ways to Split a String.go
type CharSet struct {
    chars []int
    size int
}

func NewCharSet() *CharSet {
    return &CharSet{
        chars: make([]int, 26),
        size: 0,
    }
}

func (s *CharSet) Add(char byte) {
    index := char - 'a'

    if s.chars[index] == 0 {
        s.size++
    }

    s.chars[index]++
}

func (s *CharSet) Remove(char byte) {
    index := char - 'a'

    if s.chars[index] == 0 {
        return
    }

    if s.chars[index] == 1 {
        s.size--
    }

    s.chars[index]--
}

func (s *CharSet) Size() int {
    return s.size
}


func numSplits(s string) int {
    n := len(s)
    leftSet, rightSet := NewCharSet(), NewCharSet()

    for i := 0; i < n; i++ {
        rightSet.Add(s[i])
    }

    count := 0
    for i := 0; i < n; i++ {
        rightSet.Remove(s[i])
        leftSet.Add(s[i])

        if leftSet.Size() == rightSet.Size() {
            count++
        }
    }

    return count
}

File: 1525 Number of Good Ways to Split a String/1525 Number of Good Ways to Split a String.py
"""
You are given a string s.
A split is called good if you can split s into two non-empty strings sleft and sright where their concatenation is equal to s (i.e., sleft + sright = s) and the number of distinct letters in sleft and sright is the same.
Return the number of good splits you can make in s.

Example 1:

Input: s = "aacaba"
Output: 2
Explanation: There are 5 ways to split "aacaba" and 2 of them are good.
("a", "acaba") Left string and right string contains 1 and 3 different letters respectively.
("aa", "caba") Left string and right string contains 1 and 3 different letters respectively.
("aac", "aba") Left string and right string contains 2 and 2 different letters respectively (good split).
("aaca", "ba") Left string and right string contains 2 and 2 different letters respectively (good split).
("aacab", "a") Left string and right string contains 3 and 1 different letters respectively.

Example 2:

Input: s = "abcd"
Output: 1
Explanation: Split the string as follows ("ab", "cd").
"""
class Solution:
    def numSplits(self, s: str) -> int:
        # this is not neccessary, but speeds things up
        length = len(s)
        if length == 1:  # never splittable
            return 0
        elif length == 2:  # always splittable
            return 1

		# we are recording the first and last occurance of each included letter
        first = {}  # max size = 26
        last = {}  # max size = 26

        for index, character in enumerate(s):  # O(n)
            if character not in first:
                first[character] = index
            last[character] = index

		# we are concatenating the collected indices into a list and sort them
        indices = list(first.values()) + list(last.values())  # max length 26 + 26 = 52
        indices.sort()  # sorting is constant O(1) because of the length limit above

		# all possible splits will be in the middle of this list
        middle = len(indices)//2  # always an integer because indices has an even length

		# there are this many possible splits between the two 'median' numbers
        return indices[middle] - indices[middle-1]
